### DISCUSSION TITLE
[Proposal] Assigning one value to all items in a tuple

### DISCUSSION THREAD

**Discussion by AustinBryan** [dotnet/csharplang#1725] [Language Ideas]
### Problem:
It feels sort've redundant to do something like 

```csharp
(X, Y, Z) = forceCube ? (GetSizeForX(), GetSizeForY(), GetSizeForZ()) : (GetSizeForX(), GetSizeForX(), GetSizeForX())
```

Where each item is getting the same value. 

### Purposed Solution:
So what if there was a way we could write something like this:

```csharp
(X, Y, Z) = forceCube ? (GetSizeForX(), GetSizeForY(), GetSizeForZ()) : GetSizeForX();
```

And `true` will be assigned to all three variables. This makes the code shorter, easier to read, and if I decide later I want them to all start off as false, I change it in only one place.

### **Questions:**

If someone did
```csharp
(List1, List2, List3) = new List<int>();
// instead of:
(List1, List2, List3) = (new List<int>(), new List<int>(), new List<int>());
```

Do they all have their own `List<int>()` (meaning the right hand of the assignment gets executed once for each item in the tuple) or they all share the same reference to that one list? (meaning the right hand gets executed only once in total). I think it would be more useful if they each got their own `List<int>` so that way we could quickly and easily initialize multiple variables of the same time. 

**EDIT:** Made better example

**Comment by HaloFour** [comment]
```cs
public MyType() => IsFoo = HasBar = ShouldBuzz = true;
```

**Comment by AustinBryan** [comment]
@HaloFour Wow, can't believe I forgot that, I use that pretty frequently too. Talk about overthinking it 😂

**Comment by AustinBryan** [comment]
Actually, I realized there are times where it would be more uniquely useful, and that's assigning to multiple values. 

```csharp
(X, Y, Z) = forceCube ? (GetSizeForX(), GetSizeForY(), GetSizeForZ()) : (GetSizeForX(), GetSizeForX(), GetSizeForX())
(X, Y, Z) = forceCube ? (GetSizeForX(), GetSizeForY(), GetSizeForZ()) : GetSizeForX();
```
The first line doesn't even fit on this page without having to be scrolled.
The bottom line is cleaner and is less redundant. It's also much nicer than using three conditional operators, or an if/else statement.

**Comment by svick** [comment]
> The bottom line is cleaner

I'm not sure it is. Based on the names of the methods you used, I would probably expect that if the condition is `false`, only `X` is assigned.

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/issues/1456#issuecomment-380819381

**Comment by jnm2** [comment]
The bottom line gives me the impression that it's trying to smooth over something inherently arbitrary and make it look fundamental.

**Comment by theunrepentantgeek** [comment]
> The first line doesn't even fit on this page without having to be scrolled.

We don't need language changes to fix that.
```
(X, Y, Z) = forceCube 
    ? (GetSizeForX(), GetSizeForY(), GetSizeForZ()) 
    : (GetSizeForX(), GetSizeForX(), GetSizeForX());
```

Regardless of the formatting, if I saw that in a code review I'd flag it for modification - using the ternary operator obfuscates the fact that X is always assigned the same thing. (I'd also observe that the sense of `forceCube` seems to be inverted - `Y` and `Z` only end up set to `X` if `forceCube` is **false**.)

To me, this is much clearer:
```
X = GetSizeForX();
Y = forceCube ? GetSizeForX() : GetSizeForY();
Z = forceCube ? GetSizeForX() : GetSizeForZ();
```

**Comment by yaakov-h** [comment]
Except you'd be calling `GetSizeForX()` several times instead of storing the variable.

But then you could do this:

```csharp
X = GetSizeForX();
Y = forceCube ? X : GetSizeForY();
Z = forceCube ? X : GetSizeForZ();
```

**Comment by Thaina** [comment]
```C#
public static (T,T) RepeatTuple2<T>(this T obj) => (obj,obj);
public static (T,T,T) RepeatTuple3<T>(this T obj) => (obj,obj,obj);
public static (T,T,T,T) RepeatTuple4<T>(this T obj) => (obj,obj,obj,obj);

////

(X, Y, Z) = forceCube ? (GetSizeForX(), GetSizeForY(), GetSizeForZ()) : GetSizeForX().RepeatTuple3();
```
